src/group_theory/perm/sign.lean

```diff
@@ -49,6 +49,39 @@ def subtype_perm (f : perm α) {p : α → Prop} (h : ∀ x, p x ↔ p (f x)) :
   λ _, by simp only [perm.inv_apply_self, subtype.coe_eta, subtype.coe_mk],
   λ _, by simp only [perm.apply_inv_self, subtype.coe_eta, subtype.coe_mk]⟩
 
+lemma perm_inv_on_of_perm_on_finset {s : finset α} {f : perm α}
+  (h : ∀ x ∈ s, f x ∈ s) {y : α} (hy : y ∈ s) : f⁻¹ y ∈ s :=
+begin
+  have h0 : ∀ y ∈ s, ∃ x (hx : x ∈ s), y = (λ i (hi : i ∈ s), f i) x hx :=
+    finset.surj_on_of_inj_on_of_card_le (λ x hx, (λ i hi, f i) x hx)
+    (λ a ha, h a ha) (λ a₁ a₂ ha₁ ha₂ heq, (equiv.apply_eq_iff_eq f).mp heq) rfl.ge,
+  obtain ⟨y2, hy2, heq⟩ := h0 y hy,
+  convert hy2,
+  rw heq,
+  simp only [inv_apply_self]
+end
+
+lemma perm_inv_on_of_perm_on_fintype  {f : perm α} {p : α → Prop} (hf : fintype {x // p x})
+  (h : ∀ x, p x → p (f x)) {x : α} (hx : p x) : p (f⁻¹ x) :=
+begin
+  let s := @set.to_finset _ {x | p x} hf,
+  have h : ∀ a ∈ s, f a ∈ s :=
+    λ (a : α) (ha : a ∈ s), set.mem_to_finset.mpr (h a (set.mem_to_finset.mp ha)),
+  have h2 : ∀ b ∈ s, f⁻¹ b ∈ s := λ (b : α) (hb : b ∈ s), perm_inv_on_of_perm_on_finset h hb,
+  apply set.mem_to_finset.mp, apply h2,
+  refine set.mem_to_finset.mpr hx
+end
```

Comment:
To me this feels like a nicer way to express this lemma:
```lean
lemma perm_inv_maps_to_of_maps_to {f : perm α} {s : set α} [fintype s]
  (h : set.maps_to f s s) : set.maps_to (f⁻¹ : _) s s :=
λ x hx, set.mem_to_finset.mp $
  perm_inv_on_of_perm_on_finset
    (λ a ha, set.mem_to_finset.mpr (h (set.mem_to_finset.mp ha)))
    (set.mem_to_finset.mpr hx)
```

Reply:
I'll work on these...

Reply:
How on earth does lean know that f⁻¹⁻¹ = f in the ← part of the proof of perm_inv_maps_to_iff_maps_to? As far as I can tell that isn't defeq and needs at least a simp. How did you know to even try?
I'm just as surprised that I can just apply perm_inv_maps_to_of_maps_to to the p (f⁻¹ x) form (but I guess I can convince myself why). These few lines make me realize just how much further I have still to go... (and makes me doubt I'll ever get there)

Reply:
> How on earth does lean know that f⁻¹⁻¹ = f in the ← part of the proof of perm_inv_maps_to_iff_maps_to? 

Because `f⁻¹` is just `{ to_fun := f.inv_fun, inv_fun := f.to_fun, ...}`, so just by tracing the assignments lean can see the fields of `f⁻¹⁻¹` and `f` are equal.

> How did you know to even try?

I only tried that because I wanted to be reminded of the error message so that I knew what rewrite to apply, and it turned out I didn't need one at all!